azure-mgmt-containerinstance/src/main/java/com/microsoft/azure/management/containerinstance/implementation/ContainerGroupImpl.java

```diff
@@ -0,0 +1,314 @@
+                return impl;
+            }
+        });
+    }
+
+    @Override
+    protected Observable<ContainerGroupInner> getInnerAsync() {
+        return this.manager().inner().containerGroups().getByResourceGroupAsync(this.resourceGroupName(), this.name());
+    }
+
+    @Override
+    public ContainerGroupImpl withLinux() {
+        this.inner().withOsType(OperatingSystemTypes.LINUX);
+
+        return this;
+    }
+
+    @Override
+    public ContainerGroupImpl withWindows() {
+        this.inner().withOsType(OperatingSystemTypes.WINDOWS);
+
+        return this;
+    }
+
+    @Override
+    public ContainerGroupImpl withPublicImageRegistryOnly() {
+        this.inner().withImageRegistryCredentials(null);
+
+        return this;
+    }
+
+    @Override
+    public ContainerGroupImpl withPrivateImageRegistry(String server, String username, String password) {
+        if (this.inner().imageRegistryCredentials() == null) {
+            this.inner().withImageRegistryCredentials(new ArrayList<ImageRegistryCredential>());
+        }
+        this.inner().imageRegistryCredentials().add(new ImageRegistryCredential()
+            .withServer(server)
+            .withUsername(username)
+            .withPassword(password));
+
+        return this;
+    }
+
+    @Override
+    public VolumeImpl defineVolume(String name) {
+        return new VolumeImpl(this, name);
+    }
+
+    @Override
+    public ContainerGroupImpl withoutVolume() {
+        this.inner().withVolumes(null);
+
+        return this;
+    }
+
+    @Override
+    public ContainerImpl defineContainerInstance(String name) {
+        return new ContainerImpl(this, name);
+    }
+
+    @Override
+    public ContainerGroupImpl withContainerInstance(String imageName) {
+        return this.defineContainerInstance(this.name())
+            .withImage(imageName)
+            .withoutPorts()
+            .withCPUCoreCount(1)
+            .withMemorySizeInGB(1.5)
+            .attach();
+    }
+
+    @Override
+    public ContainerGroupImpl withContainerInstance(String imageName, int port) {
+        return this.defineContainerInstance(this.name())
+            .withImage(imageName)
+            .withExternalTcpPort(port)
+            .withCPUCoreCount(1)
+            .withMemorySizeInGB(1.5)
+            .attach();
+    }
+
+    @Override
+    public Map<String, Container> containers() {
+        return this.containers;
+    }
+
+    @Override
+    public Collection<Port> externalPorts() {
+        if (this.inner().ipAddress() != null && this.inner().ipAddress().ports() != null) {
+            return this.inner().ipAddress().ports();
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public int[] externalTcpPorts() {
+        return this.externalTcpPorts;
+    }
+
+    @Override
+    public int[] externalUdpPorts() {
+        return this.externalUdpPorts;
+    }
+
+    @Override
+    public Map<String, Volume> volumes() {
+        return this.volumes;
```

Comment:
related to the comments about initializing `this.volumes`, let;'s make sure this never returns null, only empty when no children.

Reply:
so the convention is to return empty collection rather than a null whenever collections are returned, right?